### PR TITLE
HPCC-12785 Output JSON BigIntegers as strings

### PR DESCRIPTION
**Pull request by afishbeck**
Requires passing size of unigned and interger ECL types to
output interfaces.

Signed-off-by: Anthony Fishbeck anthony.fishbeck@lexisnexis.com

### CODE REVIEW

File: common/fileview2/fvresultset.cpp [hunk 1]
```diff
@@ -3309,11 +3309,11 @@ extern FILEVIEW_API void writeFullWorkUnitResults(const char *username, const ch
 
             writer.outputBeginNested(getSeverityTagname(severity, flags), false);
             if (code)
-                writer.outputUInt(code, "Code");
+                writer.outputUInt(code, 0, "Code");
```
Comment (ghalliday) [inline common/fileview2/fvresultset.cpp]: Why not use 4/sizeof(unsigned)?

Comment (ghalliday) [inline common/fileview2/fvresultset.cpp]: (Although the function currently only checks if < 7, that doesn't mean another implementation may not do something else with it.)

File: common/thorhelper/roxiehelper.cpp [hunk 2]
```diff
@@ -1126,6 +1126,17 @@ void FlushingStringBuffer::startScalar(const char *resultName, unsigned sequence
     }
 }
 
+void FlushingStringBuffer::setScalarInt(const char *resultName, unsigned sequence, __int64 value, unsigned size)
+{
+    startScalar(resultName, sequence);
+    s.appendf("%"I64F"d", value);
```
Comment (ghalliday) [inline common/thorhelper/roxiehelper.cpp]: New code should have spaces between the quotes and I64F (C++11 compatibility)

Comment (ghalliday) [inline common/thorhelper/roxiehelper.cpp]: I think you can use s.append(value);

File: common/thorhelper/roxiehelper.cpp [hunk 3]
```diff
@@ -1190,6 +1201,25 @@ void FlushingJsonBuffer::startScalar(const char *resultName, unsigned sequence)
         tail.set("}]}");
     }
 }
+
+void FlushingJsonBuffer::setScalarInt(const char *resultName, unsigned sequence, __int64 value, unsigned size)
+{
+    startScalar(resultName, sequence);
+    if (size < 7)
```
Comment (ghalliday) [inline common/thorhelper/roxiehelper.cpp]: Should have a comment: // JavaSCript only supports 53 significant digits, or whatever the reason is...

File: common/thorhelper/thorxmlwrite.cpp [hunk 4]
```diff
@@ -486,16 +486,31 @@ void CommonJsonWriter::outputData(unsigned len, const void *field, const char *f
...
+        out.append('"').append(field).append('"');
+    }
 }
-
-void CommonJsonWriter::outputUInt(unsigned __int64 field, const char *fieldname)
+void CommonJsonWriter::outputUInt(unsigned __int64 field, unsigned size, const char *fieldname)
 {
-    checkDelimit();
-    appendJSONValue(out, checkItemName(fieldname), field);
+    if (size < 6)
```
Comment (ghalliday) [inline common/thorhelper/thorxmlwrite.cpp]: why < 6?

File: ecl/eclagent/eclagent.ipp [hunk 5]
```diff
@@ -443,7 +443,7 @@ public:
     void setStandAloneOptions(bool _isStandAloneExe, bool _isRemoteWorkunit, bool _resolveFilesLocally, bool _writeResultsToStdout, outputFmts _outputFmt, IUserDescriptor *_standAloneUDesc);
     inline bool needToLockWorkunit() { return !isStandAloneExe; }           //If standalone exe then either no dali, or a unique remote workunit.
 
-    virtual void setResultInt(const char * stepname, unsigned sequence, __int64);
+    virtual void setResultInt(const char * stepname, unsigned sequence, __int64, unsigned size=0);
```
Comment (ghalliday) [inline ecl/eclagent/eclagent.ipp]: I don't think I would have the default parameters

File: ecl/hqlcpp/hqlhtcpp.cpp [hunk 6]
```diff
@@ -10913,6 +10918,7 @@ void HqlCppTranslator::buildXmlSerializeScalar(BuildCtx & ctx, IHqlExpression *
     case type_swapint:
     case type_packedint:
     case type_bitfield:
+        size.setown(createConstant((__int64)type->getSize()));
```
Comment (ghalliday) [inline ecl/hqlcpp/hqlhtcpp.cpp]: getSizetConstant as above would be better

File: common/thorhelper/thorxmlwrite.cpp (hunk 4 above)
Comment (afishbeck) [inline common/thorhelper/thorxmlwrite.cpp]: JavaScript doesn't have unsigned either so wanted to compensate for the increased range.

**Comment by afishbeck** [comment]
@ghalliday Updated based on review

**Pull request merged**